File: workshop/05-langfuse/agent-traced.py
# LangChain агент з Langfuse трасуванням — Python версія
# Різниця від agent.py — тільки 3 рядки
# Запуск: python 05-langfuse/agent-traced.py

import asyncio
import os
from dotenv import load_dotenv

load_dotenv()

from langchain_mcp_adapters.client import MultiServerMCPClient
from langchain_openai import ChatOpenAI
from langgraph.prebuilt import create_react_agent
from langfuse.callback import CallbackHandler  # ← 1. Новий імпорт


async def main():
    # 2. Langfuse handler — читає ключі з .env
    langfuse_handler = CallbackHandler(
        public_key=os.environ["LANGFUSE_PUBLIC_KEY"],
        secret_key=os.environ["LANGFUSE_SECRET_KEY"],
        host=os.environ["LANGFUSE_BASEURL"],
    )

    # Все те саме що в agent.py ↓
    async with MultiServerMCPClient({
        "math": {
            "transport": "stdio",
            "command": "python",
            "args": ["./01-server/server.py"],
        },
        "autoria": {
            "transport": "stdio",
            "command": "python",
            "args": ["./02-server/server.py"],
        },
    }) as client:
        tools = await client.get_tools()

        llm = ChatOpenAI(
            api_key=os.environ["LITELLM_API_KEY"],
            base_url=os.environ["LITELLM_BASE_URL"],
            model=os.environ["LLM_MODEL"],
        )

        agent = create_react_agent(llm, tools)

        # 3. Єдина зміна: config з callbacks
        result = await agent.ainvoke(
            {"messages": [{"role": "user", "content": "Знайди BMW X5 дизель до 20000 доларів"}]},
            config={"callbacks": [langfuse_handler]}  # ← ось і все
        )

        print(result["messages"][-1].content)

    # Важливо: flush перед завершенням
    langfuse_handler.flush()


asyncio.run(main())

File: workshop/05-langfuse/agent-traced.js
// LangChain агент з Langfuse трасуванням
// Різниця від agent.js — тільки 3 рядки (імпорт + ініціалізація + callbacks)
// Запуск: npm run traced

import "dotenv/config";
import { MultiServerMCPClient } from "@langchain/mcp-adapters";
import { ChatOpenAI } from "@langchain/openai";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { CallbackHandler } from "langfuse-langchain"; // ← 1. Новий імпорт

// 2. Langfuse handler — читає ключі з .env автоматично
const langfuseHandler = new CallbackHandler({
  publicKey: process.env.LANGFUSE_PUBLIC_KEY,
  secretKey: process.env.LANGFUSE_SECRET_KEY,
  baseUrl: process.env.LANGFUSE_BASEURL,
});

// Все те саме що в agent.js ↓
const mcpClient = new MultiServerMCPClient({
  mcpServers: {
    math: {
      transport: "stdio",
      command: "node",
      args: ["./01-server/server.js"],
    },
    autoria: {
      transport: "stdio",
      command: "node",
      args: ["./02-server/server.js"],
    },
  },
});

const tools = await mcpClient.getTools();

const llm = new ChatOpenAI({
  apiKey: process.env.LITELLM_API_KEY,
  configuration: { baseURL: process.env.LITELLM_BASE_URL },
  model: process.env.LLM_MODEL,
});

const agent = createReactAgent({ llm, tools });

// 3. Єдина зміна при виклику — callbacks: [langfuseHandler]
const result = await agent.invoke(
  { messages: [{ role: "user", content: "Порахуй за допомогою інструмента add 34+12+56" }] },
  { callbacks: [langfuseHandler] }  // ← ось і все трасування
);

console.log(result.messages.at(-1).content);

// Важливо: flush надсилає дані в Langfuse перед завершенням процесу
await langfuseHandler.flushAsync();
await mcpClient.close();
